http/server_sent_event.ts

```diff
@@ -0,0 +1,411 @@
+   * (30 seconds). If set to a number, then it polls that number of
+   * milliseconds.  The feature is disabled if set to `false`.  It defaults to
+   * `false`. */
+  keepAlive?: boolean | number;
+}
+
+class CloseEvent extends Event {
+  constructor(eventInit: EventInit) {
+    super("close", eventInit);
+  }
+}
+
+/** An event which contains information which will be sent to the remote
+ * connection and be made available in an `EventSource` as an event. A server
+ * creates new events and dispatches them on the target which will then be
+ * sent to a client.
+ *
+ * See more about Server-sent events on [MDN](https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events)
+ *
+ * ### Example
+ *
+ * ```ts
+ * import {
+ *   ServerSentEvent,
+ *   SSEStreamTarget,
+ * } from "https://deno.land/std@$STD_VERSION/http/server_sent_event.ts";
+ * import { serve } from "https://deno.land/std@$STD_VERSION/http/server.ts";
+ *
+ * await serve((request) => {
+ *   const target = new SSEStreamTarget();
+ *   const evt = new ServerSentEvent(
+ *     "message",
+ *     { hello: "world" },
+ *     { id: 1 },
+ *   );
+ *   target.dispatchEvent(evt);
+ *   return target.asResponse();
+ * }, { port: 8000 });
+ * ```
+ */
+export class ServerSentEvent extends Event {
+  #data: string;
+  #id?: number;
+  #type: string;
+
+  /**
+   * @param type the event type that will be available on the client. The type
+   *             of `"message"` will be handled specifically as a message
+   *             server-side event.
+   * @param data  arbitrary data to send to the client, data this is a string
+   *              will be sent unmodified, otherwise `JSON.parse()` will be used
+   *              to serialize the value
+   * @param eventInit initialization options for the event
+   */
+  constructor(
+    type: string,
+    data: unknown,
+    eventInit: ServerSentEventInit = {},
+  ) {
+    super(type, eventInit);
+    const { replacer, space } = eventInit;
+    this.#type = type;
+    try {
+      this.#data = typeof data === "string"
+        ? data
+        : JSON.stringify(data, replacer as (string | number)[], space);
+    } catch (e) {
+      assert(e instanceof Error);
+      throw new TypeError(
+        `data could not be coerced into a serialized string.\n  ${e.message}`,
+      );
+    }
+    const { id } = eventInit;
+    this.#id = id;
+  }
+
+  /** The data associated with the event, which will be sent to the client and
+   * be made available in the `EventSource`. */
+  get data(): string {
+    return this.#data;
+  }
+
+  /** The optional ID associated with the event that will be sent to the client
+   * and be made available in the `EventSource`. */
+  get id(): number | undefined {
+    return this.#id;
+  }
+
+  override toString(): string {
+    const data = `data: ${this.#data.split("\n").join("\ndata: ")}\n`;
+    return `${this.#type === "__message" ? "" : `event: ${this.#type}\n`}${
```

Comment:
`__message` type is used for representing unnamed event type, but not documented. It also feels a little artificial. How about using empty string for representing unnamed event type?

Reply:
No, it is a sigil to identify a "message" versus an event, which gets called by a different handler on the client side.  It isn't documented because it is "private" and only used by `.dispatchMessage()`.  I will convert it to a symbol though to avoid any confusion.

Reply:
Actually tried that and the problem is that you can't call `super()` on the `Event` with that, and we need some sort of string sigil that identifies the non-event message and a `""` feels like a risky one.  Again it isn't "undocumented" it is just not intended to be publically accessible, therefore something that is unlikely to collide with a legitimate event type.